src/gui/dialogs/file_dialog.cpp

```diff
@@ -0,0 +1,512 @@
+			// FIXME: Perhaps redundant since the three-params call to normalize_path()
+			//        above necessarily validates existence.
+			DBG_FILEDLG << "register_new_selection(): new selection '" << name << "' is a file, symbolic link, or special: " << new_path << '\n';
+			current_dir_ = fs::directory_name(new_path);
+			current_entry_ = fs::base_name(new_path);
+			return SELECTION_IS_FILE;
+		}
+	}
+
+	// The path does not exist, at least not entirely. See if the parent does
+	// (in save mode non-existent files are accepted as long as the parent dir
+	// exists).
+	const std::string& absolute_parent = fs::normalize_path(new_parent, true, true);
+	if(!absolute_parent.empty()) {
+		DBG_FILEDLG << "register_new_selection(): new selection '" << name << "' does not exist or is not accessible, but parent exists\n";
+		current_dir_ = absolute_parent;
+		current_entry_ = fs::base_name(name);
+		return SELECTION_NOT_FOUND;
+	}
+
+	DBG_FILEDLG << "register_new_selection(): new selection '" << name << "' does not exist or is not accessible\n";
+	return SELECTION_PARENT_NOT_FOUND;
+}
+
+void tfile_dialog::refresh_fileview(twindow& window)
+{
+	cursor::setter cur{cursor::WAIT};
+
+	dir_files_.clear();
+	dir_subdirs_.clear();
+
+	// TODO: Need to detect and handle cases where we don't have search permission
+	//       on current_dir_, otherwise things may get weird.
+	filesystem::get_files_in_dir(current_dir_, &dir_files_, &dir_subdirs_, filesystem::FILE_NAME_ONLY);
+
+	//
+	// Clear and refill the filelist box.
+	//
+
+	tlistbox& filelist = find_widget<tlistbox>(&window, "filelist", false);
+
+	filelist.clear();
+
+	// Parent entry
+	if(!fs::is_root(current_dir_)) {
+		// label_parent may not necessarily be always ".." in the future, so push
+		// with check_selection = false and check the selection ourselves here.
+		push_fileview_row(filelist, label_parent, icon_parent, false);
+		if(current_entry_ == PARENT_DIR || current_entry_.empty()) {
+			filelist.select_row(0, true);
+		}
+	}
+
+	for(const auto& dir : dir_subdirs_) {
+		push_fileview_row(filelist, dir, icon_dir);
+	}
+
+	for(const auto& file : dir_files_) {
+		push_fileview_row(filelist, file, icon_file);
+	}
+
+	find_widget<tcontrol>(&window, "current_dir", false).set_label(current_dir_);
+	find_widget<ttext_box>(&window, "filename", false).set_value(current_entry_);
+}
+
+void tfile_dialog::push_fileview_row(tlistbox& filelist, const std::string& name, const std::string& icon, bool check_selection)
+{
+	// TODO: Hopefully some day GUI2 will allow us to make labels be ellipsized
+	//       dynamically at layout/rendering time.
+	std::string label = name;
+	utils::ellipsis_truncate(label, FILE_DIALOG_MAX_ENTRY_LENGTH);
+
+	std::map<std::string, string_map> data;
+	data["icon"]["label"] = icon;
+	data["file"]["label"] = label;
+	filelist.add_row(data);
+
+	const unsigned last_pos = filelist.get_item_count() - 1;
+	tgrid* const last_grid = filelist.get_row_grid(last_pos);
```

Comment:
I made add_row return the added row, meaning you can replace this with `tgrid* last_grid = &filelist.add_row(data); `.